types/jest/index.d.ts

```diff
@@ -757,8 +758,25 @@ declare namespace jest {
         toMatch(expected: string | RegExp): R;
         /**
          * Used to check that a JavaScript object matches a subset of the properties of an object
+         *
+         * Optionally, you can provide an object to use as Generic type for the expected value.
+         * This ensures that the matching object matches the structure of the provided object-like type.
+         *
+         * @example
+         *
+         * type House = {
+         *   bath: boolean;
+         *   bedrooms: number;
+         *   kitchen: {
+         *     amenities: string[];
+         *     area: number;
+         *     wallColor: string;
+         *   }
+         * };
+         *
+         * expect(desiredHouse).toMatchObject<House>(...standardHouse, kitchen: {area: 20}) // wherein standardHouse is some base object of type House
          */
-        toMatchObject(expected: {} | any[]): R;
+        toMatchObject<O extends {}>(expected: O | any[]): R;
```

Comment:
`toMatchObject` can also work with arrays:

> You can also pass an array of objects, in which case the method will return true only if each object in the received array matches (in the toMatchObject sense described above) the corresponding object in the expected array.

So we could probably support arrays in the generic as well. For example, currently the following will pass without any issues.

```ts
.toMatchObject<[{ foo: string }, { bar: number }]>([{ foo: true }, { bar: () => {} }]); // no errors!
```

What do you think about something like this? This should add type safety for tuples and array types via the generic if desired.

```suggestion
        toMatchObject<E extends {} | any[]>(expected: E): R;
```

Reply:
That's a very good suggestion. I've made your changes as well as added a corresponding test in dbfa03d852c22dbe723f559a8029ce389278570a